src/compiler_internals.md

```diff
@@ -0,0 +1,37 @@
+
+**JS:** *AMM* looks surprisingly close to Javascript/Typescript, so a transpiler was written that mostly just strips out the type information and replaces all of opcode calls with method calls on [a pretty simple runtime shim](https://github.com/alantech/alan/blob/main/js-runtime/index.js).
+
+The transpiled Javascript doesn't have any of the benefits of the language in terms of cpu parallelism or automatic concurrency, but with some extra intelligence in the transpilation and judicious use of WebWorker threads that could be restored.
+
+**AGA** looks very similar to assembly, but with annotation on events, closures, and the dependency graph that the Alan Runtime uses reorganize for concurrency and parallelism benefits.
+
+**AGC** is the binary version of *AGA* and is a near direct translation of the text to a 64-bit little-endian integer packed array, with keywords being given numeric values that are clearly interpretable as 8 UTF-8/ASCII characters.
+
+The compiler contains as a set of AST generators for the various input formats ([ln](https://github.com/alantech/alan/tree/main/compiler/src/ln), [amm](https://github.com/alantech/alan/blob/main/compiler/src/amm.ts), [aga](https://github.com/alantech/alan/blob/main/compiler/src/aga.ts)) with ln using [ANTLR](https://www.antlr.org/) to define its grammar and the other two using our homegrown [lp](https://github.com/alantech/alan/blob/main/compiler/src/lp.ts).
+
+Next, the compiler has the translation stages that go from one input format to another output format ([lntoamm](https://github.com/alantech/alan/tree/main/compiler/src/lntoamm), [ammtojs](https://github.com/alantech/alan/blob/main/compiler/src/ammtojs.ts), [ammtoaga](https://github.com/alantech/alan/blob/main/compiler/src/ammtoaga.ts), [agatoagc](https://github.com/alantech/alan/blob/main/compiler/src/agatoagc.ts)).
+
+While overkill at the moment, the compiler [solves the shortest path from the specified input file to output file and constructs the pipeline to do so on the fly](https://github.com/alantech/alan/blob/main/compiler/src/pipeline.ts). (This is much faster than it sounds and has allowed us to add new inputs/outputs and translation formats quickly.)
+
+The first stage of the compiler dominates with more than 90% of the total compile time, and within that, the ANTLR-generated parser takes over 99% of the time within the `lntoamm` layer. (Speeding this up is very important to us.) Once the AST is generated, the first layer transforms this into a scope-oriented collection of elements and then generates AMM code on a per-event-handler basis, finally it looks for constants in these handlers and hoists them to the global scope and deduplicates them, then emits this in the AMM format.
```

Comment:
let's title these paragraphs: LN to AMM, AMM to JS, AMM to AGA, AGA to AGC ?

Reply:
Sounds good. I did think it felt like a wall of text but I wasn't sure how to adjust it.